x-pack/plugins/alerting/server/task_runner/maintenance_windows/maintenance_windows_service.ts

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { KibanaRequest, Logger } from '@kbn/core/server';
+import { MaintenanceWindow } from '../../application/maintenance_window/types';
+import { filterMaintenanceWindowsIds } from './get_maintenance_windows';
+import { MaintenanceWindowClientApi } from '../../types';
+import { AlertingEventLogger } from '../../lib/alerting_event_logger/alerting_event_logger';
+import { withAlertingSpan } from '../lib';
+
+export const DEFAULT_CACHE_INTERVAL_MS = 60000; // 1 minute cache
+
+interface MaintenanceWindowServiceOpts {
+  cacheInterval?: number;
+  getMaintenanceWindowClientWithRequest(request: KibanaRequest): MaintenanceWindowClientApi;
+  logger: Logger;
+}
+
+interface MaintenanceWindowData {
+  maintenanceWindows: MaintenanceWindow[];
+  maintenanceWindowsWithoutScopedQueryIds: string[];
+}
+
+interface GetMaintenanceWindowsOpts {
+  request: KibanaRequest;
+  spaceId: string;
+}
+
+type LoadMaintenanceWindowsOpts = GetMaintenanceWindowsOpts & {
+  eventLogger: AlertingEventLogger;
+  ruleTypeCategory: string;
+};
+
+interface LastUpdatedWindows {
+  lastUpdated: number;
+  activeMaintenanceWindows: MaintenanceWindow[];
+}
+
+export class MaintenanceWindowsService {
+  private cacheIntervalMs = DEFAULT_CACHE_INTERVAL_MS;
+
+  private windows: Map<string, LastUpdatedWindows> = new Map();
+
+  constructor(private readonly options: MaintenanceWindowServiceOpts) {
+    if (options.cacheInterval) {
+      this.cacheIntervalMs = options.cacheInterval;
+    }
+  }
+
+  public async loadMaintenanceWindows(
+    opts: LoadMaintenanceWindowsOpts
+  ): Promise<MaintenanceWindowData> {
+    const activeMaintenanceWindows = await this.getMaintenanceWindows({
+      request: opts.request,
+      spaceId: opts.spaceId,
+    });
+
+    // Only look at maintenance windows for this rule category
+    const maintenanceWindows = activeMaintenanceWindows.filter(({ categoryIds }) => {
```

Comment:
active maintenance windows are cached but whether those active maintenance windows apply to this rule type based on category are checked every time this function is called.